### DISCUSSION TITLE
How to handle table updates with referential constraints

### DISCUSSION THREAD

**Discussion by arco400** [IBM/ibmi-tobi#351] [Q&A]
@edmundreinhardt 

What would be the best way to handle updates for parent tables that have referential constraints defined to them?

When I try to execute a CREATE OR REPLACE of the parent table, the process stops because it cannot replace the primary key that ensures the unicity.
I tried the process with changing the referential constraints to disabled state, but that does not help.
Probably I need to remove the referential constraints before the CREATE OR REPLACE statement is executed. And recreate the referential constraints after finishing the process. 
If I go that way, what would be the recommended place to define these actions so that BOB understands to execute these actions before and after the table change?

**Comment by AIforIBMi** [comment]
If these actions to remove referential integrity and add them are SQL statements then just put them into the .table file where each SQL statement will be executed in order.
Does that work?

**Comment by arco400** [comment]
That might work for incremental builds, but when doing a complete build the removal may fail because they are not yet there, and adding of the constraints may also fail because the child table is not yet there.
Is there a recommended way to circumvent this?

**Reply by AIforIBMi** [reply]
yes, there is a conditional section that we have a PR for that will only run if there is an existing object.  @irfanshar will be eventually merging this and then we can address this use case

**Comment by edmundreinhardt** [comment]
Another solution is for us to have commands that can fail, and to swallow those errors.
This is already supported today
You just have to prefix with !
See https://github.com/IBM/bob-recursive-example/blob/9a11c819b443c1c33ba4139b6a9234993e2f337a/QBNDSRC/SAMPLE.BNDDIR#L4

**Comment by arco400** [comment]
The thing with referential constraints is that they are not program objects like triggers, so testing for their existance might be an issue. Or do you mean to test for existance of the table object?

**Comment by AIforIBMi** [comment]
yes, I meant the existence of the table itself

**Reply by arco400** [reply]
It would be nice to have that conditional section. Thanks!

**Comment by arco400** [comment]
When thinking it over, I see some issues with placing the ADD CONSTRAINT statements in either the table build script for the parent table or the child table. For adding referential constraints it is always required that both tables exist. And I do not want to duplicate constraint definitions on multiple places.

Maybe not an ideal solution, but is it possible to configure BOB so that we can run a pre-buiild-project and a post-build-project script per build process where the removal and adding of the constraints could be fired? In the documentation I see a mention of option installProjectScript in iproj.json. Is that meant to do something similar?
